asyncnsq/tcp/writer.py

```diff
@@ -182,9 +196,39 @@ async def mpub(self, topic, *messages):
     def id(self):
         return self._conn.endpoint
 
-    def close(self):
+    async def cancel(self, timeout = 10):
+        time_in = time.time()
+        self._is_working = False
         self._conn.close()
+        while not self._conn.closed:
+            await asyncio.sleep(0.1, loop=self._loop)
+            now = time.time()
+            if timeout > 0 and now - time_in > timeout:
+                logger.warning("close failed: timeout")
+        logger.info("closing _auto_reconnect_task")
         self._status = consts.CLOSED
+        self._auto_reconnect_task.cancel()
+        while not self._auto_reconnect_task_closed:
+            await asyncio.sleep(0.1, loop=self._loop)
+            now = time.time()
+            if timeout > 0 and now - time_in > timeout:
+                logger.warning("cancel failed: timeout")
+        logger.info("_auto_reconnect_task_closed = {}".format(self._auto_reconnect_task_closed))
+
+    def close(self, timeout = 10):
```

Comment:
why close function  needs a loop?

Reply:
it is a blocking non-async function. it is waiting of the async function really closed. We can also use a queue to receive the signal, but that could make the implementation of `timeout` even more ugly.

Reply:
a loop blocking function will have further issues with async loop.
maybe we can find a bertter way to do it.

Reply:
close is a non-async function. we can call `cancel`.

Actually if we can use `async close()`, the `cancel` function did everything we need. However, it may change the origin API.

If people still using `writer.close()` in async function, they will got a runtime error

```
RuntimeWarning: coroutine 'Reader.close' was never awaited
```

I was afraid this breaking change is not acceptable.

Reply:
yes, breaking change is not acceptable for now.

we need figure out a more elegant way to do it, or we can add a plus method with this feature can called manuallly without having breaking change, if you need this feature.

what do you think?

Reply:
since actually `cancel` did everything `close` should done. I believe `cancel` is that plus method.

Beside, since the origin api writer.close didn't really close every connections, actually it will not cause additionally damage.